### ISSUE TITLE
Expired todos cannot be recreated with --update-todo if expired todo file exists

### ISSUE THREAD

**Issue by scalvert** [ember-template-lint/ember-template-lint#1905]
The issue here is a combination of conditions that compose together to create a situation where you cannot create a new todo, and are not aware that a todo file exists.

Steps to reproduce:

1. Create a todo for an error in a file, set an expiry date
2. Let it expire
3. Do not run `--fix` (which removes the stale todo file)
4. Running `ember-template-lint <filePath>` shows an error (the todo is expired)
5. Running `ember-template-lint <filePath> --update-todo` will _not_ create a new todo file, since the file already exists (it's in the `stable` map that's returned in the batches
6. There's no error message about stale todos because it's in the `stable` map (we only push the error when it's in the `remove` map)

Should we be pushing an error related to expired todos similar to how we do stale?

cc/ @rwjblue @MelSumner

**Comment by MelSumner** [comment]
I don't think this is a bug; this is how we intended it to work. The idea is that the todo has entered the error state and should now be fixed, because the time has run out to casually "fix it when you get to it".  

We could do a flag like `--refresh-todos` (I had even previously suggested this), but I wonder if it might be against the whole idea of having the todos in the first place. 

They can already delete the todo files and re-create, but especially in cases where individual rules are set for specific warn/error dates (thinking about a11y-related rules here), the user who does this is likely to use the generic command (`yarn lint:hbs --update-todo`) and miss the point entirely of having customized warn/error dates.

**Comment by scalvert** [comment]
It's not really how we intended it to work, since you can no longer know that there's an active todo; the file is essentially orphaned until the error is fixed, at which point we'd push the message to delete the stale todo file.

I think this is a case we need to handle better, as it feels like the feature is 'broken'.

**Comment by MelSumner** [comment]
I think the explicit error state because the todo is "expired" is something that can provide the motivation to fix the linting issue  . The friction of going out of your way to figure out how to delete the todo and re-create it is hopefully the catalyst for improvement. 

If we do want to offer this, I do still think it should be a separate flag/command.

**Comment by scalvert** [comment]
This is definitely a bug, and has nothing to do with workflows. If the todo due dates have expired, this file serves no purpose anymore and should just be deleted. This is state is akin to not having used todos at all in the first place.

I think we need to auto-delete all expired todo files. That will fix this issue.

**Comment by MelSumner** [comment]
Ok. I guess I rationalized it as an artifact that the todo _did_ exist, but has been left ignored and entered into an error state. 

For reporting purposes, I'd like to be able to keep track of when the issue was first created as a todo. I suppose that could be the responsibility of the reporting tool and not the linter, though.

**Comment by MelSumner** [comment]
Update: we're going to check for expired todos and create a `--clean-todo` flag.

**Comment by MelSumner** [comment]
@scalvert @rwjblue what do we want to do about this: 

`1 errors and 0 warnings potentially fixable with the \`--fix\` option.`

Since we want to

1. keep removing expired todos in the `--fix` option 
2. want to add `--clean-todo` to explicitly only remove expired todos
 
what should this message be? 

I think our options are: 

1. keep the message the same 
2. add `--clean-todo` so it reads  `1 errors and 0 warnings potentially fixable with the \`--fix\` or \'--clean-todo\' option.`
3. separate the message out so we report expired todos and instruct running the `--clean-todo` option
4. something else?

**Comment by scalvert** [comment]
> keep removing expired todos in the --fix option

Expired todos should really be a completely opaque operation from the perspective of developers. This seems like no change there?

> want to add --clean-todo to explicitly only remove expired todos

We want `--clean-todo` to remove stale todos _and_ expired todos. 

I think the message can stay the same.

**Comment by MelSumner** [comment]
Right, expired = stale and invalid. 👍 

Ok, I will keep the message the same. Thanks for the clarification!

**Comment by scalvert** [comment]
Well expired is literally that - the errorDate is less than today. Stale is that the todo was fixed, or the lines/columns drifted (today, thought we want to fix that).

**Comment by MelSumner** [comment]
Ok I think I was just using different terms so let me make sure I understand: 

- expired = the errorDate has passed
- stale = the todo was fixed or is no longer valid because the code line was moved or file was moved

So then, `--clean-todo` will remove both of these things. The `--fix` flag will do autofixes and also remove both of these things. 

Is that correct?

**Comment by scalvert** [comment]
So I believe we're _not_ going to fix stale todos with `--fix`, but only with `--clean-todo`, right? That was the whole point, otherwise why have both?

**Comment by scalvert** [comment]
Said another way:

`--clean-todo` - removes stale and expired todos
`--fix` - removes only expired todos

**Comment by MelSumner** [comment]
> So I believe we're _not_ going to fix stale todos with `--fix`, but only with `--clean-todo`, right? That was the whole point, otherwise why have both?

I thought @rwjblue had the requirement that `--fix` should do all the things and should still be the go-to that folks become trained to use, but agreed that we could add `--clean-todos` to remove expired and stale todos but NOT autofix anything else.

**Comment by scalvert** [comment]
I don't think there's any 'requirement'. What moving off `--fix` as the primary thing that does the modification of todos does for us is removes unintended changes to files that happen via that command. Having an explicitly different command helps with that.

**Comment by MelSumner** [comment]
Ok so based on that, I think we have unintentionally ended up with a message that might confuse users: 
Working through [updating this test](https://github.com/ember-template-lint/ember-template-lint/blob/a2f58d3a45cd051815729af7b28f8cc491df1ba0/test/acceptance/todo-cli-test.js#L323), we'd have this message: 

```
        "app/templates/require-button-type.hbs
          -:-  error  Todo violation passes \`require-button-type\` rule. Please run \`ember-template-lint app/templates/require-button-type.hbs --clean-todo\` to remove this todo from the todo list.  invalid-todo-violation-rule

        ✖ 1 problems (1 errors, 0 warnings)
          1 errors and 0 warnings potentially fixable with the \`--fix\` option."
```

**Comment by scalvert** [comment]
How is that confusing? The todo message is explicitly for todos. Are you saying it's confusing with the fix message below?

**Comment by scalvert** [comment]
Discussing this a bit with @rwjblue, we both agree that what you said is correct (though I initially leaned towards `--fix` doing less):

`--clean-todo` - removes stale and expired todos
`--fix` - fixes fixables _and_ removes stale and expired todos

The `--clean-todo` flag is a subset operation of `--fix`.

**Comment by scalvert** [comment]
Resolved in #1933
